File: backend/app/schemas/task.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

TaskStatus = Literal["todo", "in_progress", "done"]
TaskPriority = Literal["high", "medium", "low"]


class TaskCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    status: TaskStatus = "todo"
    priority: TaskPriority = "medium"
    due_date: date | None = None


class TaskUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    status: TaskStatus | None = None
    priority: TaskPriority | None = None
    due_date: date | None = None


class TaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str | None
    status: str
    priority: str
    due_date: date | None
    created_at: datetime
    updated_at: datetime

File: frontend/src/api/task.ts
import client from './client'

export type TaskStatus = 'todo' | 'in_progress' | 'done'
export type TaskPriority = 'high' | 'medium' | 'low'

export interface Task {
  id: number
  title: string
  description: string | null
  status: TaskStatus
  priority: TaskPriority
  due_date: string | null
  created_at: string
  updated_at: string
}

export interface TaskPayload {
  title: string
  description?: string
  status?: TaskStatus
  priority?: TaskPriority
  due_date?: string
}

export async function listTasks(params?: {
  status?: TaskStatus
  priority?: TaskPriority
}): Promise<Task[]> {
  return (await client.get('/tasks', { params })) as unknown as Promise<Task[]>
}

export async function createTask(payload: TaskPayload): Promise<Task> {
  return (await client.post('/tasks', payload)) as unknown as Promise<Task>
}

export async function updateTask(id: number, payload: Partial<TaskPayload>): Promise<Task> {
  return (await client.patch(`/tasks/${id}`, payload)) as unknown as Promise<Task>
}

export async function deleteTask(id: number): Promise<void> {
  await client.delete(`/tasks/${id}`)
}
